### DISCUSSION TITLE
NPC Records, Nullability (ocassional lack thereof) and change detection.

### DISCUSSION THREAD

**Discussion by MadKillerChicken** [Mutagen-Modding/Mutagen#433] [Q&A] [answered]
Fist of all let me thank you for this awesome project, found it recently and I love it so much.

My current intent is to try to implement some sort of "bundled patching" of everything NPC related (Appearance, AI, Inventory, so on).
I started to experiment a bit and was wondering about some of the AI related fields.
It seems impossible to determine if a field, like `AIData`, contains data or simply has "default" values.
Most field come with the convenient `IsNull` property, but some not.

Do I need to compare all values in such fields with the data from the "master" file to determine if a change has occurred?

Thanks again!

**Comment by Noggog** [comment, marked as answer]
> It seems impossible to determine if a field, like `AIData`, contains data or simply has "default" values.
Most field come with the convenient `IsNull` property, but some not.

Nullability is not meant to indicate records that are deviating from the default values (or the previous values).   Instead, they're meant mostly for communicating whether the data exists when stored on disk or not
https://github.com/Mutagen-Modding/Mutagen/wiki/Nullability-to-Indicate-Record-Presence
But this is unrelated to whether RecordA in OverrideMod.esp has a value that is different from how it looked in BaseMaster.esp

> Do I need to compare all values in such fields with the data from the "master" file to determine if a change has occurred?

Yes, to determine what has changed from the previous override, you would need to look at and compare all field's content and see what changed.  You might want to lean on the Equality concepts, like GetEqualsMask.   They haven't been fully tested and confirmed correct, though https://github.com/Mutagen-Modding/Mutagen/issues/211
But it might be better to push that over the finish line rather than reinventing the wheel.

However, the equality systems Mutagen will provide with its built in Equality and GetEqualsMask is going to be a gritty interpretation:  Eg, -any- difference, however minute, will mark it as not equal.  So if you have a more subjective interpretation of what fields "matter" and how, then you might need to implement your own equality checks with those subjective bits in mind

Lastly, if you find yourself writing, by hand, hundreds of lines of verbose equality code for every field of every object, then I would suggest considering some alternatives such as reflection, code generation, or some other means of avoiding hand writing tons of error prone code like that.

**Reply by MadKillerChicken** [reply]
Thank you very much for such a detailed answer.

While I've been consuming TES mods for a long time, I'm quite new to actually creating them.
So I'll need to first understand the structure of the records I intend to deal with and how exactly the games uses them. Avoiding huge amounts of error prone equality code is still far in the future.

**Reply by Noggog** [reply]
Just so you're aware, Mutagen exposes a lot of the records in strongly typed fashions:
`npc.Health = 15;` 
and such.

These are super nice for when you know what things you want to modify and how, as most Synthesis patchers are doing.   They care about modifying fields A, B, C, and that's it, so writing lines that modify ABC is nice and easy.

However, when you get into the realm of modifying everything in a super generic fashion, it's less appropriate to write code that modifies fields directly, as I've been mentioning.   There's another "style" of code that's needed for hyper generic logic, which won't actually ever touch specific fields on specific records directly.   These styles are usually more advanced, but doable.  Definitely feel free to ask questions if that's the route you're going down, either here, or on discord

**Reply by MadKillerChicken** [reply]
Thanks a lot for you kindness and willingness to help out, it's very much appreciated.
I'll surely will post more questions when needed 👍
